Handle 'P' operator (#3)

IndianBoy42 opened the pull request:
a bit of code duplication but they are short functions anyway.

Adds the paste keymaps depending on which one is available

IndianBoy42 committed:
handle paste_behind

AckslD commented on lua/anywise_reg/handlers.lua:
```diff
@@ -18,6 +18,17 @@ M.handle_action = function(prefix, operator, textobject)
     data.update_reg_data(reg_name, operator, textobject)
 end
 
+M.handle_paste_behind = function(prefix)
```
I think it would be better to change `handle_paste` to take an `operator` argument (which will be `p` or `P`) and avoid some code duplication.

AckslD commented on lua/anywise_reg/keybinds.lua:
```diff
@@ -54,22 +54,30 @@ local function setup_paste_keymaps(prefix)
-local function setup_keymaps_for_prefix(prefix)
-    setup_yank_keymaps(prefix)
-    setup_paste_keymaps(prefix)
+local function setup_paste_behind_keymaps(prefix)
```
Same comment as for the handler, I think this could take in a `operator` argument

AckslD commented on lua/anywise_reg/keybinds.lua:
```diff
@@ -54,22 +54,30 @@ local function setup_paste_keymaps(prefix)
         end
+        for c = 97, 122 do table.insert(prefixes, '"' .. string.char(c)) end
+
+        for _, prefix in ipairs(prefixes) do setup_yank_keymaps(prefix) end
```
Doesn't this now happen only it either paste keys are `nil`?

I think we only need one loop over the `prefixes` and inside that one check if the paste keys should be setup

IndianBoy42 replied:
I think the condition means not both nil? AKA at least one of them has a value

I can change to one loop though no prpblem

AckslD commented on lua/anywise_reg/config.lua:
```diff
@@ -4,6 +4,7 @@ M.config = {
     operators = {},
     textobjects = {},
     paste_key = nil,
+    paste_behind_key = nil,
```
Could you add this to the README as well?

AckslD reviewed (changes requested):
Thanks @IndianBoy42! I left some comments

IndianBoy42 committed:
Dedup and generalize code

handle_past and setup_paste_keymaps take an operator.
setup_paste_keymaps also takes a the key
the user config is changed to take a `table paste_keys`
that maps keys to operators,
so technically other paste operators than 'p' and 'P' could be used

IndianBoy42 commented:
The code is more DRY now, and the readme is updated, I also generalized the config to 'config.paste_keys' which takes a table, please check if it is clear how that works

AckslD commented on lua/anywise_reg/keybinds.lua:
```diff
@@ -48,36 +48,26 @@ local function setup_yank_keymaps(prefix)
-            prefix .. '\')<CR>'
+local function setup_paste_keymaps(prefix, key, operator)
+    local lhs = prefix .. key
+    local rhs = '<Cmd>lua require("anywise_reg.handlers").handle_paste(\'' ..
+                    prefix .. '\',\'' .. operator .. '\')<CR>'
```
You could use `format_str_args` here, see https://github.com/AckslD/nvim-anywise-reg.lua/blob/main/lua/anywise_reg/keybinds.lua#L45

AckslD commented on lua/anywise_reg/keybinds.lua:
```diff
@@ -48,36 +48,26 @@ local function setup_yank_keymaps(prefix)
     local prefixes = {'', '""'}
-    if config.paste_key ~= nil or config.paste_behind_key ~= nil then
+    if #config.paste_keys >= 1 then
+        -- if config.paste_key ~= nil or config.paste_behind_key ~= nil then
```
This comment can be removed

AckslD commented on lua/anywise_reg/keybinds.lua:
```diff
@@ -48,36 +48,26 @@ local function setup_yank_keymaps(prefix)
 M.setup_keymaps = function()
     local prefixes = {'', '""'}
-    if config.paste_key ~= nil or config.paste_behind_key ~= nil then
+    if #config.paste_keys >= 1 then
```
I think we should still register the action keymaps even if the user doesn't want the paste ones. The user might set the paste keys themselves for some reason

IndianBoy42 replied:
Ah I just based that on your repo has 

```
    if config.paste_key ~= nil then
```

AckslD replied:
Ah yeah, I think that was wrong :)

AckslD commented on lua/anywise_reg/handlers.lua:
```diff
@@ -29,15 +29,15 @@ M.handle_paste_behind = function(prefix)
     if d ~= nil then
         -- go to end of current text object
-        cmd.normal('v'..d.textobject, {noremap = false})
+        cmd.normal('v' .. d.textobject, {noremap = false})
```
I guess there should be an if-statement here to check for `P` and if so do `o`

IndianBoy42 replied:
Oh sorry I forgot about this, do you think it is possible there are other paste operators? that would need different handling of this section? so should we make it configurable?

AckslD replied:
No problem, I think for now it's fine to assume that it's only `p` and `P`. If there will be others we can change that then :)

AckslD reviewed (changes requested):
Looks good @IndianBoy42! A few more comments

IndianBoy42 committed:
cleanup of the code

AckslD commented:
Thanks for the update @IndianBoy42, one last thing probably is if you could change to use 4 spaces throughout instead of tabs to be consistent.

IndianBoy42 commented:
> Thanks for the update @IndianBoy42, one last thing probably is if you could change to use 4 spaces throughout instead of tabs to be consistent.

Hmm the easiest way to do this is running stylua/luaformat, but both of those introduce a lot more changes in your code so till now I had just manually excluded those from the commits

I can fix it but what do you think about adopting stylua? (it is being considered for use by the neovim project)

IndianBoy42 committed:
Fix indentation

AckslD commented:
> Hmm the easiest way to do this is running stylua/luaformat, but both of those introduce a lot more changes in your code so till now I had just manually excluded those from the commits
> 
> I can fix it but what do you think about adopting stylua? (it is being considered for use by the neovim project)

That might be good, although I haven't used it yet. Anyway, I'd prefer formatting changes to be part of a separate PR :)

For now could you do for example
```vim
:set ts=8 sw=4 sts=4
gg=G
```
?

IndianBoy42 commented:
The indentation should be correct in the latest commit

AckslD commented:
Just saw thanks! I added a deprecation warning for `paste_key` in the config such that it suddenly doesn't break if people are already using it :+1: Otherwise it's merged now
